csrc/scheduler/matmul_heuristic_plugin.cpp

```diff
@@ -0,0 +1,287 @@
+      .cta_tile =
+          {(uint16_t)params.tile_sizes.cta_tile.m,
+           (uint16_t)params.tile_sizes.cta_tile.n,
+           (uint16_t)params.tile_sizes.cta_tile.k},
+      .warp_tile =
+          {(uint16_t)params.tile_sizes.warp_tile.m,
+           (uint16_t)params.tile_sizes.warp_tile.n,
+           (uint16_t)params.tile_sizes.warp_tile.k},
+      .instruction_tile =
+          {(uint16_t)params.tile_sizes.instruction_tile.m,
+           (uint16_t)params.tile_sizes.instruction_tile.n,
+           (uint16_t)params.tile_sizes.instruction_tile.k},
+      .splitk_factor = (uint16_t)params.splitk_factor,
+      .load_stages =
+          (uint8_t)params.double_buffer_options.smem_double_buffer_stage,
+      .grid_swizzle_factor = (uint8_t)params.grid_swizzle_factor,
+      .cta_order = (uint8_t)toUnderlying(params.cta_order),
+      .double_buffer_smem_read =
+          params.double_buffer_options.double_buffer_smem_read,
+      .rotate_ldmatrix_out_of_main_loop =
+          params.rotate_ldmatrix_out_of_main_loop,
+  };
+  plugin.getConfig(&config, &problem);
+
+  const auto setTile = [](GemmTile& gemm_tile,
+                          const KernelConfig::Tile& input) {
+    gemm_tile.m = input[0];
+    gemm_tile.n = input[1];
+    gemm_tile.k = input[2];
+  };
+
+  params.double_buffer_options.smem_double_buffer_stage = config.load_stages;
+  setTile(params.tile_sizes.cta_tile, config.cta_tile);
+  setTile(params.tile_sizes.warp_tile, config.warp_tile);
+  setTile(params.tile_sizes.instruction_tile, config.instruction_tile);
+
+  // Update mma macro if necessary to match instruction tile
+  MmaMacroEncode menc(params.mma_macro); // this will record the family
+  menc.m = config.instruction_tile[0]; // update instruction tile size
+  menc.n = config.instruction_tile[1];
+  menc.k = config.instruction_tile[2];
+  params.mma_macro = menc; // cast back to uint64_t
+
+  params.splitk_factor = config.splitk_factor;
+  params.grid_swizzle_factor = config.grid_swizzle_factor;
+  switch (config.cta_order) {
+    case 0:
+      params.cta_order = MatmulParams::TileRasterizationOrder::RowMajor;
+      break;
+    case 1:
+      params.cta_order = MatmulParams::TileRasterizationOrder::ColumnMajor;
+      break;
+    default:
+      NVF_ERROR(
+          false,
+          "Unrecognized cta_order returned by plugin: ",
+          config.cta_order,
+          ". Expected 0 (row-major) or 1 (column-major)");
+  }
+  params.double_buffer_options.double_buffer_smem_read =
+      config.double_buffer_smem_read;
+  params.rotate_ldmatrix_out_of_main_loop =
+      config.rotate_ldmatrix_out_of_main_loop;
+
+  // enable double buffering or circular buffering if configured
+  params.double_buffer_options.double_buffer_smem_write =
+      config.load_stages > 1;
+
+  // async load only for circular buffering (stages > 2)
+  params.async_gmem_load_operands = config.load_stages > 2;
+
+  return true;
+}
+
+} // namespace matmul_heuristic_plugin
```

Comment:
Shouldn't all this belong to the plugin? That is, nvfuser does not assume the structure of `KernelConfig` and `ProblemDescription`, it just treat them as opaque data. The plugin should provide API to create and free these objects, and get/set data of these objects, and nvFuser just call these APIs without knowing anything about the internal representation of such objects.

Reply:
Good point. I think there's a way to address this along with your other comment about code duplication. We could get rid of all the getters and setters by just declaring a struct with a virtual `configure` method in the api header. Then the plugin should implement a subclass of that class and implement `configure`. That can be private to the plugin i.e. opaque to nvfuser; they only need to export a factory function that creates one of these objects. I think that will simplify the nvfuser code quite a bit without burdening the plugin developers.